edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/stream/StreamRequestListener.cs

```diff
@@ -0,0 +1,207 @@
+        {
+            this.streamRequestHandlerProvider = Preconditions.CheckNotNull(streamRequestHandlerProvider, nameof(streamRequestHandlerProvider));
+            this.streamLock = new SemaphoreSlim(maxConcurrentStreams);
+        }
+
+        public void InitPump(IModuleClient moduleClient)
+        {
+            Events.StartingPump();
+            this.pumpTask = this.Pump(moduleClient);
+        }
+
+        public void Dispose()
+        {
+            this.cts.Cancel();
+            Events.StoppingPump();
+            // Don't dispose the cts here as it can cause an ObjectDisposedException.
+        }
+
+        async Task Pump(IModuleClient moduleClient)
+        {
+            while (!this.cts.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.streamLock.WaitAsync(this.cts.Token);
+                    await this.ProcessStreamRequest(moduleClient);
+                }
+                catch (Exception e)
+                {
+                    Events.ErrorInPump(e);
+                }
+            }
+        }
+
+        async Task ProcessStreamRequest(IModuleClient moduleClient)
+        {
+            Option<(string requestName, IClientWebSocket clientWebSocket)> result = await this.WaitForStreamRequest(moduleClient);
+            // If we don't get a valid stream, release the lock.
+            if (!result.HasValue)
+            {
+                this.streamLock.Release();
+            }
+            else
+            {
+                // We got a WebSocket stream. Pass it to a handler on a background thread. This thread is
+                // responsible for releasing the lock after it completes.
+                result.ForEach(
+                    r =>
+                    {
+                        Events.NewStreamRequest(r.requestName, this.streamLock.CurrentCount);
+                        this.HandleRequest(r.requestName, r.clientWebSocket);
+                    });
+            }
+        }
+
+        async Task<Option<(string requestName, IClientWebSocket clientWebSocket)>> WaitForStreamRequest(IModuleClient moduleClient)
+        {
+            var result = Option.None<(string, IClientWebSocket)>();
+            try
+            {
+                DeviceStreamRequest deviceStreamRequest = await moduleClient.WaitForDeviceStreamRequestAsync(this.cts.Token);
+                if (deviceStreamRequest != null)
+                {
+                    IClientWebSocket clientWebSocket = await moduleClient.AcceptDeviceStreamingRequestAndConnect(deviceStreamRequest, this.cts.Token);
+                    result = Option.Some((deviceStreamRequest.Name, clientWebSocket));
+                }
+            }
+            catch (Exception ex)
+            {
+                Events.ErrorGettingStream(ex);
+            }
+
+            return result;
+        }
+
+        async void HandleRequest(string requestName, IClientWebSocket clientWebSocket)
```

Comment:
Does this need to be an `async void`?

Reply:
Emm, any other pattern that starts a background thread without using a library like Hangfire, such as Task.Run seemed basically equivalent. 
I didn't want to get into keeping track of all the threads and checking their status, etc.